tests/test_utils.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2017 "Shopify inc." All rights reserved.
+# Use of this source code is governed by a MIT-style license that can be found in the LICENSE file.
+from __future__ import unicode_literals
+
+import xml.etree.ElementTree as et
+
+import pytest
+import tests.utils
+
+
+@pytest.fixture
+def valid_workflow():
+    return """
+<workflow-app xmlns="uri:oozie:workflow:0.5" name="descriptive-name">
+    <parameters>
+        <property>
+            <name>property_key</name>
+            <value>property_value</value>
+        </property>
+        <property>
+            <name>other_property_key</name>
+            <value>🤣</value>
+        </property>
+    </parameters>
+    <global>
+        <job-tracker>job-tracker</job-tracker>
+        <name-node>name-node</name-node>
+    </global>
+    <start to="end" />
+    <end name="end" />
+</workflow-app>
+""".strip()
+
+
+def test_xml_to_dict_unordered():
+    document_dict = tests.utils.xml_to_dict_unordered("""
+<root>
+  <tag />
+  <tag key="value" />
+  <tag>Text</tag>
+</root>
+    """.strip())
+    assert document_dict == {
+        'root': {
+            'tag': [None, u'Text', {'@key': 'value'}]
+        }
+    }
+
+
+def test_assert_valid_workflow(valid_workflow):
+    # A minimal workflow should validate
+    tests.utils.assert_valid_workflow(valid_workflow)
+
+    # With valid XML, duplicate names should raise an assertion error
+    with pytest.raises(AssertionError) as assertion_info:
+        xml = """
+<workflow-app xmlns="uri:oozie:workflow:0.5" name="descriptive-name">
+    <action name="alpha"/>
+    <action name="alpha"/>
+    <action name="beta"/>
+</workflow-app>""".strip()
+        assert len(et.fromstring(xml)), 'Invalid test XML'
+        tests.utils.assert_valid_workflow(xml)
+    assert str(assertion_info.value) == 'Name(s) reused: alpha'
+
+    # With valid XML, a missing a start tag should result in a schema violation assertion error
+    with pytest.raises(AssertionError) as assertion_info:
+        xml = """
+<workflow-app xmlns="uri:oozie:workflow:0.5" name="descriptive-name">
+    <global>
+        <job-tracker>job-tracker</job-tracker>
+        <name-node>name-node</name-node>
+    </global>
+    <end name="end" />
+</workflow-app>""".strip()
+        assert len(et.fromstring(xml)), 'Invalid test XML'
+        tests.utils.assert_valid_workflow(xml)
+    assert str(assertion_info.value).strip() == str(
+        '''An XML validation error occurred.
+
+Error: Invalid app definition, org.xml.sax.SAXParseException; lineNumber: 6; columnNumber: 23; '''
```

Comment:
If SAXParser changes its code or we update its version, it may happen that this exact error may not remain valid.
Is it possible to shrink this error with the most relevant search string which you think might stay across versions ?